zerver/lib/idempotent_request.py

```diff
@@ -0,0 +1,180 @@
+            )
+            SELECT new_id FROM insertion_cte AS new_id
+        """
+        ).format(
+            dest_table=Identifier(destination_model._meta.db_table),
+            idempotency_table=Identifier(IdempotentRequest._meta.db_table),
+            cols=SQL(", ").join(map(Identifier, field_names)),
+            values_placeholders=SQL(", ").join(map(Placeholder, field_names)),
+            pk=Identifier(destination_model._meta.pk.name),
+        )
+        cursor.execute(
+            query,
+            {
+                **column_to_value,
+                "idempotency_key_value": idempotency_key,
+                "realm_id": realm.id,
+                "user_id": user.id,
+            },
+        )
+        result = cursor.fetchall()
+
+        # We should only get a single result row.
+        assert len(result) == 1
+        return result[0]
+
+
+def idempotent_request_transaction(
+    durable: bool = False,
+    savepoint: bool = True,
+    *,
+    cached_response_serializer: Callable[[Any], dict[str, Any]],
+    cached_response_deserializer: Callable[[str], Any],
+) -> Callable[[Callable[ParamT, WorkReturnT]], Callable[ParamT, WorkReturnT]]:
+    """
+    Our custom decorator to apply idempotency for
+    non-idempotent operations (e.g. do_send_messages)
+    and to more finely control the start/end of a transaction.
+
+    This decorator expects the decorated function to receive
+    idempotency_key and user as kwargs, Both are optional but are required
+    to apply idempotency.
+
+    Note: Currently only wraps do_send_messages().
+    """
+
+    def idempotency_decorator(
+        do_work: Callable[ParamT, WorkReturnT],
+    ) -> Callable[ParamT, WorkReturnT]:
+        @wraps(do_work)
+        def wrapper(*args: ParamT.args, **kwargs: ParamT.kwargs) -> WorkReturnT:
+            # Cast it to a dict to access its values like a normal dict and keep mypy happy.
+            # **kwargs (without ParamSpec) is already treated as dict in python anyway.
+            _kwargs = cast(dict[str, Any], kwargs)
+            idempotency_key = _kwargs.get("idempotency_key")
+            user = _kwargs.get("user")
+
+            # Idempotency-Key is optional, so we should still do the work normally
+            # if i'ts omitted by client.
+            if idempotency_key is None:
+                with transaction.atomic(durable=durable, savepoint=savepoint):
+                    return do_work(*args, **kwargs)
+
+            assert user is not None
+            with connection.connection.cursor(cursor_factory=DictCursor) as cursor:
+                query = SQL("""
+                            INSERT INTO {idempotency_table} (realm_id, user_id, idempotency_key)
+                            VALUES (%(realm_id)s, %(user_id)s, %(idempotency_key_value)s)
+                            ON CONFLICT DO NOTHING;
+
+                            SELECT cached_response AS cached_response FROM {idempotency_table}
```

Comment:
I don't understand the need to say "cached_response AS cached_response", but maybe that's just my ignorance of SQL.

Reply:
I know of no reason to have an explicit `AS` here.